pwndbg/gdblib/kernel/slab.py

```diff
@@ -35,3 +45,256 @@ def oo_order(x: int) -> int:
 
 def oo_objects(x: int) -> int:
     return int(x) & OO_MASK
+
+
+_flags = {
+    "SLAB_DEBUG_FREE": 0x00000100,
+    "SLAB_RED_ZONE": 0x00000400,
+    "SLAB_POISON": 0x00000800,
+    "SLAB_HWCACHE_ALIGN": 0x00002000,
+    "SLAB_CACHE_DMA": 0x00004000,
+    "SLAB_STORE_USER": 0x00010000,
+    "SLAB_RECLAIM_ACCOUNT": 0x00020000,
+    "SLAB_PANIC": 0x00040000,
+    "SLAB_DESTROY_BY_RCU": 0x00080000,
+    "SLAB_MEM_SPREAD": 0x00100000,
+    "SLAB_TRACE": 0x00200000,
+    "SLAB_DEBUG_OBJECTS": 0x00400000,
+    "SLAB_NOLEAKTRACE": 0x00800000,
+    "SLAB_NOTRACK": 0x01000000,
+    "SLAB_FAILSLAB": 0x02000000,
+}
+
+
+def get_flags_list(flags: int) -> List[str]:
+    return [flag_name for flag_name, mask in _flags.items() if flags & mask]
+
+
+class Freelist:
+    def __init__(self, start_addr: int, offset: int, random: int = 0):
+        self.start_addr = start_addr
+        self.offset = offset
+        self.random = random
+
+    def __iter__(self) -> Generator[int, None, None]:
+        current_object = self.start_addr
+        while current_object:
+            addr = int(current_object)
+            yield current_object
+            current_object = memory.pvoid(addr + self.offset)
+            if self.random:
+                current_object ^= self.random ^ swab(addr + self.offset)
+
+    def __int__(self) -> int:
+        return self.start_addr
+
+    def __len__(self) -> int:
+        return sum(1 for _ in self)
+
+    def find_next(self, addr: int) -> int:
+        freelist_iter = iter(self)
+        for obj in freelist_iter:
+            if obj == addr:
+                return next(freelist_iter, 0)
+        return 0
+
+
+class SlabCache:
+    def __init__(self, slab_cache: gdb.Value):
+        self._slab_cache = slab_cache
+
+    @property
+    def address(self) -> int:
+        return int(self._slab_cache)
+
+    @property
+    def name(self) -> str:
+        return self._slab_cache["name"].string()
+
+    @property
+    def offset(self) -> int:
+        return int(self._slab_cache["offset"])
+
+    @property
+    def random(self) -> int:
+        return (
+            int(self._slab_cache["random"])
+            if kernel.kconfig().get("SLAB_FREELIST_HARDENED") == "y"
```

Comment:
I think we can do `if "SLAB_FREELIST_HARDENED" in kernel.kconfig()`, which at least to me looks nicer.

Reply:
This would work in the qemu test images we use, but might cause errors in other configurations. What if someone is debugging a system with `CONFIG_SLAB_FREELIST_HARDENED=n`? Then the `if` condition would be true, because that config is set, but once the logic would start "derandomizing" it would actually start to scramble things up.

Correct me if I'm wrong here...

Reply:
> I'm not even sure if the =n syntax is valid

Right, it's not valid. `# CONFIG_<optionName> is not set` is how you specify that an option is not set [^1]. For boolean options, I feel like checking if it's equal to "y" is the same as `if some_bool == True` instead of `if some_bool`, which feels strange to me. Checking the value would make more sense for any non-boolean options.

[^1]: https://stackoverflow.com/questions/41946817/which-approach-to-use-for-kernel-config

Reply:
ok, I changed it now